tensortrade/stochastic/processes/gan.py

```diff
@@ -0,0 +1,381 @@
+            rnn_outputs, rnn_states = tf.nn.dynamic_rnn(
+                cell=cell,
+                dtype=tf.float32,
+                inputs=decoder_input)
+            rnn_outputs_flat = tf.reshape(rnn_outputs, [-1, self._hidden_units_g])
+            logits = tf.matmul(rnn_outputs_flat, self._W_out_D) + self._b_out_D
+            output = tf.nn.sigmoid(logits)
+        return output, logits
+
+    def price_gen(self, ind_gen_sample=1):
+        daily_log_returns = self._generated_data[0, :, ind_gen_sample]
+
+        price_list = [self._training_data['Close'][-1]]  # initial price (today)
+
+        for x in np.transpose(daily_log_returns):
+            price_list.append(price_list[-1] * np.exp(x))
+
+        return price_list
+
+    def price_real(self):
+        daily_returns = np.random.normal(self._cagr / self._t, self._volatility / np.sqrt(self._t), self._t) + 1
+
+        price_list = [self._training_data['Close'][-1]]
+
+        for x in daily_returns:
+            price_list.append(price_list[-1] * x)
+
+        return price_list
+
+    def prices_gen_data_frame(self):
+        df = pd.DataFrame([])
+
+        for i in range(self._n_samples):
+            df[i] = self.price_real()
+
+        return df
+
+    def prices_real_data_frame(self):
+        df = pd.DataFrame()
+
+        for i in range(self._n_samples):
+            df[i] = self.price_real()
+
+        return df
+
+    def train_generator(self, batch_idx, offset):
+        # update the generator
+        for g in range(self._g_rounds):
+            Y_mb = self.get_batch(self._samples, self._batch_size, batch_idx + g + offset)
+            _, G_loss_curr = self._sess.run([self._G_solver, self._G_loss],
+                                            feed_dict={self._CG: Y_mb,
+                                                       self._Z: self.sample_Z(self._batch_size, self._seq_length,
+                                                                              self._latent_dim)})
+        return G_loss_curr
+
+    def train_discriminator(self, batch_idx, offset):
+        # update the discriminator
+        for d in range(self._d_rounds):
+            # using same input sequence for both the synthetic data and the real one,
+            # probably it is not a good idea...
+            X_mb = self.get_batch(self._samples, self._batch_size, batch_idx + d + offset)
+            X_mb = X_mb.reshape(self._batch_size, self._seq_length, 1)
+            Y_mb = self.get_batch(self._samples, self._batch_size, batch_idx + d + offset)
+            _, D_loss_curr = self._sess.run([self._D_solver, self._D_loss],
+                                            feed_dict={self._CD: Y_mb, self._CG: Y_mb, self._X: X_mb,
+                                                       self._Z: self.sample_Z(self._batch_size, self._seq_length,
+                                                                              self._latent_dim)})
+        return D_loss_curr
+
+    def _initialize_gan(self):
```

Comment:
This training needs to be simplified because it is written with tensorflow 1 and there has been much improvement in tensorflow 2.0 to simplify all of this.